src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/MemoryScope.java

```diff
@@ -325,6 +332,20 @@ public CleanupAction dup() {
                 return new DupAction(this);
             }
 
+            @Override
+            public CleanupAction wrap(Runnable runnable) {
+                disable();
+                return AtMostOnceOnly.of(() -> {
+                    try {
+                        runnable.run();
+                    } catch (Throwable t) {
```

Comment:
Maybe the exception should be re-thrown after doing critical cleanup?

Reply:
We could do that - but I'd prefer to leave this as is.